fix: clone the contact when creating all-contacts addressBook

kxxt opened the pull request:
We should create a clone of the contact when creating the all-contacts address book because we intend to update them separately. It will become a footgun that is hard to debug if the virtual addressbook and real addressbook is referencing the same contact object.

kxxt committed:
fix: clone the contact when creating all-contacts addressBook

jobisoft commented:
Should we not instead try to use the references correctly? I guess adding and deleting contacts can stay as it is, but updating needs a second thought? It will double the data stored if we add clones.

kxxt commented:
> Should we not instead try to use the references correctly? I guess adding and deleting contacts can stay as it is, but updating needs a second thought? It will double the data stored if we add clones.


If the contact object is updated in the real address book first. Then when our code updates the virtual address book, the same contact object will get updated twice. So we need to ensure that our functions such as `addContactToCategoryInCache` works without bugs(IIRC some of these functions rely on the categories field in contact object) even if the categories field stored in the contact is already updated. We need to document this in the code. Feel free to choose one way you like.

jobisoft committed:
Revert "fix: clone the contact when creating all-contacts addressBook"

This reverts commit 9a88e71167c3e71ac61bf2c029e01b21a2700ac6.

jobisoft committed:
check id of virtual ab instead of comparing objects

jobisoft committed:
add the same contact object to single address book and virtual all contacts address book

jobisoft committed:
fix: virtual ab was not sorted on contact modifications

jobisoft committed:
adjust code to cater for references contacts

- unify createContactInCache, modifyContactInCache and deleteContactInCache to automatically handle the virtual address book, even though there is no special handling needed for creatContactInCache (it could simply be called twice)
- reduce the number of operations by striping categories first
- make sure we do not loop over live objects, if we modify them

jobisoft committed:
wrap logging so it can be disabled

jobisoft committed:
addCategoryToContact -> updateContactCategoryCache

When a contact is modified all his cached categories must be updated, either because the categories changed or because the displayName/email changed. The function addCategoryToCachedContact is already doing the right thing, but it needs to be called on any update, not just on additions.

jobisoft committed:
expandImplicitCategories when parsing contacts

When deleting a contact we require that the cached contact has all implicit categories stored. If we only have "A / B" stored, but not "A", than category "A" will not be removed and still hold a reference to the deleted contact.

jobisoft committed:
vcard optimizations

- categories is a multi value field
- keep add/remove logic out of the vcard update function
- do not update a card twice on move

jobisoft committed:
remove unnecessary category expansion

jobisoft committed:
category fixes und unifications

- suppress empty categories
- use vCard in function names which act on the vCard.
- removeCategory() is almost identical to  moveCategory() -> replaceCategoryInVCards()
- use common merge/strip category functions

jobisoft committed:
use an actual value for the uncategorized dataset and  add option to remove contact from current category
